packages/protect/src/schema/index.ts

```diff
@@ -0,0 +1,243 @@
+export type EncryptConfig = z.infer<typeof encryptConfigSchema>
+
+// ------------------------
+// Interface definitions
+// ------------------------
+export class ProtectColumn {
+  private columnName: string
+  private castAsValue: CastAs
+  private indexesValue: {
+    ore?: OreIndexOpts
+    unique?: UniqueIndexOpts
+    match?: Required<MatchIndexOpts>
+    ste_vec?: SteVecIndexOpts
+  } = {}
+
+  constructor(columnName: string) {
+    this.columnName = columnName
+    this.castAsValue = 'text'
+  }
+
+  /**
+   * Set or override the cast_as value.
+   */
+  dataType(castAs: CastAs) {
+    this.castAsValue = castAs
+    return this
+  }
+
+  /**
+   * Enable ORE indexing (Order-Revealing Encryption).
+   */
+  orderAndSort() {
+    this.indexesValue.ore = {}
+    return this
+  }
+
+  /**
+   * Enable a Unique index. Optionally pass tokenFilters.
+   */
+  equality(tokenFilters?: TokenFilter[]) {
+    this.indexesValue.unique = {
+      token_filters: tokenFilters ?? [],
+    }
+    return this
+  }
+
+  /**
+   * Enable a Match index. Allows passing of custom match options.
+   */
+  freeTextSearch(opts?: MatchIndexOpts) {
+    // Provide defaults
+    this.indexesValue.match = {
+      tokenizer: opts?.tokenizer ?? { kind: 'ngram', token_length: 3 },
+      token_filters: opts?.token_filters ?? [
+        {
+          kind: 'downcase',
+        },
+      ],
+      k: opts?.k ?? 6,
+      m: opts?.m ?? 2048,
+      include_original: opts?.include_original ?? true,
+    }
+    return this
+  }
+
+  /**
+   * Enable a STE Vec index, requires a prefix.
+   * TODO: Determine a better way to describe ste_vec
+   */
+  steVec(prefix: string) {
+    this.indexesValue.ste_vec = { prefix }
+    return this
+  }
+
+  build() {
+    return {
+      cast_as: this.castAsValue,
+      indexes: this.indexesValue,
+    }
+  }
+
+  getName() {
+    return this.columnName
+  }
+}
+
+interface TableDefinition {
+  tableName: string
+  columns: Record<string, ColumnSchema>
+}
+
+export class ProtectTable<T extends ProtectTableColumn> {
+  constructor(
+    public readonly tableName: string,
+    private readonly columnBuilders: T,
+  ) {}
+
+  /**
+   * Build a TableDefinition object: tableName + built column configs.
+   */
+  build(): TableDefinition {
+    const builtColumns: Record<string, ColumnSchema> = {}
+    for (const [colName, builder] of Object.entries(this.columnBuilders)) {
+      builtColumns[colName] = builder.build()
+    }
+
+    return {
+      tableName: this.tableName,
+      columns: builtColumns,
+    }
+  }
+}
+
+// ------------------------
+// User facing functions
+// ------------------------
+export function csTable<T extends ProtectTableColumn>(
+  tableName: string,
+  columns: T,
+): ProtectTable<T> & T {
+  const tableBuilder = new ProtectTable(tableName, columns) as ProtectTable<T> &
+    T
+
+  for (const [colName, colBuilder] of Object.entries(columns)) {
+    ;(tableBuilder as ProtectTableColumn)[colName] = colBuilder
```

Comment:
Is the `;` here intentional?

Reply:
Our code formatting through biome adds this automatically because the previous line that didn’t end with a semicolon, and the next line begins with an open parenthesis. It just protects against bugs in code formatting